lib/3scale/backend/stats/partition_generator_job.rb

```diff
@@ -1,12 +1,39 @@
 module ThreeScale
   module Backend
     module Stats
+      # Job for deleting service stats
+      # Maps delete job definition to a set of non overlapping key set partitions
       class PartitionGeneratorJob < BackgroundJob
         # low priority queue
-        @queue = :main
+        @queue = :stats
 
         class << self
-          def perform_logged(_enqueue_time, service_id, applications, metrics, users, from, to, context_info = {})  end
+          include Configurable
+
+          def perform_logged(_enqueue_time, service_id, applications, metrics, users,
+                             from, to, context_info = {})
+            job = DeleteJobDef.new(
+              service_id: service_id,
+              applications: applications,
+              metrics: metrics,
+              users: users,
+              from: from,
+              to: to
+            )
+
+            stats_key_gen = KeyGenerator.new(job.to_hash)
+
+            # Generate partitions
+            0.step(stats_key_gen.keys.count, configuration.stats.delete_partition_batch_size).each do |idx|
```

Comment:
If this fails in some call to `Resque.enqueue`, I think that the job will be marked as failed, re-enqueued at some point, and then it will start scheduling duplicated `PartitionEraserJob`s. Is that right?

Reply:
totally right! 

happy to hear improvements

Reply:
I don't see an easy solution that fits the current approach.

You could rescue from Redis errors and retry. That could work if there's minimal downtime and retrying a few times work. If after retrying a few times the problem persists, we'd need to mark the job as failed.

What if we added an optional index to the `job` so it remembers what's already been enqueued? Would that work?

Reply:
you mean, if fails, rescue it and add `idx` to this job, so when requeued from failed jobs can resume from it? is that possible?

Reply:
I think so. I see 2 possible approaches:

- Set the index in the job and let it fail so the failed jobs scheduler picks it. We'd need to make sure that when rescheduling the failed job, Resque picks the mutated one instead of the original one.
- Set the index in the job, discard the current job, and enqueue the new one normally instead of letting it fail.

I don't have a strong preference, but I think the first approach might be better because it will allow  us to see some activity in the `failed` queue and notice that there might be something wrong with our infra.

Reply:
But if it fails, it will be by network issue. So jobs could not be re-scheduled in a "short" term.

I mean, the only error in this `PartitionGeneratorJob` can be originated enqueuing jobs. So if fails, how can be re-scheduled?

Reply:
If there's a long downtime the job will simply fail and never be retried unless we implement something like retries with back-offs. I don't think it's worth it right now.

Reply:
I think if this job fails, handing off to the `Resque::Failure` will also probably fail.

Anyway, I do not see any way to "update" de job with some index.

if `perform_logged` looks like this

```ruby
def perform_logged(_enqueue_time, service_id, applications, metrics, users,
                   from, to, partition_idx, context_info = {})
```
and I update `partition_idx`,
```ruby
partition_idx = partition_idx + 1
```

On re-scheduling, `partition_idx` will have the new value???

Reply:
I don't know. I'll need to check the code that marks the job as failed and stores it in the failed queue.

If that does not work, we always have the option of discarding the job and enqueueing a new one in the 'stats' queue instead of the failed one with Resque.enqueue. However, as you said, this might be a bit overkill, because it only protects against small network hiccups. Not sure if we should tackle this for a first version. You can give it a try and see whether it complicates the code a lot.

Reply:
tried above, updating `partition_idx` and does not work.

On calling `Resque::Job#fail` instance method,  job is handed off to the `Resque::Failure`, but we do not have access to the job_params.

Unless, anybody knows a way to handle this, I suggest we go on and write an issue to implement this kind of "smart retry" strategy.

Reply:
OK :+1:

Reply:
composing a unique key based on params and the parent job ID could be a possible solution here? :wave: 

Duplicate keys shouldn't fail so the exception should be handled gracefully. But on the other side, there could be some kind of logging in place (most probably with a small retention period) in order to be able to act upon such occurrences ie if too frequent